FIX #87: Option to suppress alerts on webpages.

danishjafri88 opened the pull request:
Added the `suppress alert` logic for WKUIDelegate methods to show prompts. This is a fix for #87 

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:
- https://dustiest-limitation.000webhostapp.com/alert.html
- https://dustiest-limitation.000webhostapp.com/prompt.html
- https://dustiest-limitation.000webhostapp.com/confirm.html
- https://dustiest-limitation.000webhostapp.com/iframe_alert.html
- https://dustiest-limitation.000webhostapp.com/iframe_prompt.html
- https://dustiest-limitation.000webhostapp.com/iframe_confirm.html
<!-- Any useful notes for reviewer explaining how best to test and verify. -->

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adaquate test plan exists for QA to validate (if applicable)

danishjafri88 committed:
Added basic logic, edge cases pending

danishjafri88 committed:
Reset block alert bool on domain change

danishjafri88 committed:
Added Localised strings and a confirmation action sheet

danishjafri88 commented:
@anthonypkeane Need copy for the alert to suppress Alert; currently:
Title: "Suppress Alerts"
Message: "Suppress Alerts on this page for this session?"

anthonypkeane commented:
@tomlowenthal can you approve (or edit) this string as its security related.

Message: Prevent this page from creating additional alerts
Suppress Alerts
Cancel

danishjafri88 committed:
Added copy for alert message

jhreis commented on Client/Frontend/Browser/BrowserPrompts.swift:
```diff
@@ -112,6 +127,11 @@ struct TextInputAlert: JSAlertInfo {
         alertController.addAction(UIAlertAction(title: Strings.OKString, style: .default) { _ in
             self.completionHandler(input.text)
         })
+        if suppressHandler != nil {
```
We really should find a way to abstract this, rather than having this logic pasted amongst three separate classes.  Even going as far as creating a new class `JSAlert: JSAlertInfo` that handles this, and having these three classes subclass that, rather than conforming to `JSAlertInfo` .

Otherwise a more simple option maybe to use a `UIAlertAction` on these classes instead of a `supressHandler`.

```
if let alertAction = alertAction {
  alertController.addAction(alertAction)
}
```

And the `runJavaScriptTextInputPanelWithPrompt` function would set an alert action with title and callback and everything, instead of each of these configuring it manually.

jhreis commented on Client/Frontend/Browser/Tab.swift:
```diff
@@ -43,7 +43,8 @@ struct TabState {
 class Tab: NSObject {
     var id: String?
-    
+    var alertShownCount: Int = 0
```
wdyt about putting these in a struct together, since they are tied together?

danishjafri88 replied:
they are actually independent of each other. So keeping them as is

jhreis replied:
I just meant since they are both related to alerts? But this is fine by me.

jhreis commented on Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:
```diff
@@ -196,6 +196,11 @@ extension BrowserViewController: WKNavigationDelegate {
                 webView.configuration.preferences.javaScriptEnabled = !domainForShields.isShieldExpected(.NoScript)
             }
+            // Reset the block alert bool on new host. 
+            if let newHost: String = url.host, let oldHost: String = webView.url?.host, newHost != oldHost {
```
No need to unwrap all of those, since we're doing equality checks:

`if url.host != webview.url?.host {`

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,59 @@ extension BrowserViewController: WKUIDelegate {
+            return
+        }
+        promptingTab.alertShownCount += 1
+        alert.suppressHandler = promptingTab.alertShownCount > 1 ? ({[unowned self] suppress in
```
I think using a ternary operator here is too easy to miss, given how large just one chunk is.  I think using a traditional conditional would be better, or at least setting the closure to a variable and making the ternary much cleaner:

`alert.suppressHandler = promptingTab.alertShownCount > 1 ? suppressAlert : nil`

jhreis commented on BraveShared/BraveStrings.swift:
```diff
@@ -88,6 +88,8 @@ public extension Strings {
     public static let OpenPhoneSettingsActionTitle = NSLocalizedString("OpenPhoneSettingsActionTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Open Settings", comment: "See http://mzl.la/1G7uHo7")
     public static let CopyImageActionTitle = NSLocalizedString("CopyImageActionTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Copy Image", comment: "Context menu item for copying an image to the clipboard")
     public static let CloseAllTabsTitle = NSLocalizedString("CloseAllTabsTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Close All %i Tabs", comment: "")
+    public static let SuppressAlertsActionTitle = NSLocalizedString("SuppressAlertsActionTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Suppress Alerts", comment: "")
```
We should add `comment` here.

jhreis commented on BraveShared/BraveStrings.swift:
```diff
@@ -88,6 +88,8 @@ public extension Strings {
     public static let CopyImageActionTitle = NSLocalizedString("CopyImageActionTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Copy Image", comment: "Context menu item for copying an image to the clipboard")
     public static let CloseAllTabsTitle = NSLocalizedString("CloseAllTabsTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Close All %i Tabs", comment: "")
+    public static let SuppressAlertsActionTitle = NSLocalizedString("SuppressAlertsActionTitle", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Suppress Alerts", comment: "")
+    public static let SuppressAlertsActionMessage = NSLocalizedString("SuppressAlertsActionMessage", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Prevent this page from creating additional alerts", comment: "")
```
We should add `comment` here.

danishjafri88 committed:
refactored the browser prompt code to remove duplicate code

jhreis commented on Client/Frontend/Browser/BrowserPrompts.swift:
```diff
@@ -15,8 +15,31 @@ import Shared
+    var info: JSAlertInfo!
+    var showsCancel: Bool = false
+    
+    class func newController(title: String?, message: String?, preferredStyle: UIAlertController.Style = .alert, info: JSAlertInfo, showCancel: Bool = true) -> JSPromptAlertController {
```
Can we make this a convenience initializer instead?

```
    convenience init(title: String?, message: String?, preferredStyle: UIAlertController.Style = .alert, info: JSAlertInfo, showCancel: Bool = true) {
        self.init(title: title, message: message, preferredStyle: preferredStyle)
        self.info = info
        self.showsCancel = showCancel
    }
```

danishjafri88 replied:
Thats not possible since, `init(title: title, message: message, preferredStyle: preferredStyle)` is a convenience initalizer itself. And the only init we can override is ` override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?)` which if used requires all params  to be nil and the super init call to be for the convenience initalizer

jhreis replied:
Convenience init, can call a convenience init.

danishjafri88 replied:
Yeah just realised, thanks!

jhreis commented on Client/Frontend/Browser/BrowserPrompts.swift:
```diff
@@ -15,8 +15,31 @@ import Shared
-    var alertInfo: JSAlertInfo?
-
+    
+    var info: JSAlertInfo!
```
Understand the reason here, but we should not force unwrap this, using an optional here is pretty easy, and is a better convention for non views.

jhreis commented on Client/Frontend/Browser/BrowserPrompts.swift:
```diff
@@ -15,8 +15,31 @@ import Shared
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        if info.suppressHandler != nil {
```
May as well unwrap this:

```
        if let handler = info?.suppressHandler {
            self.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .default, handler: { _ in
                handler(true)
            }))
        }
```

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,60 @@ extension BrowserViewController: WKUIDelegate {
+                }
+                // Show confirm alert here.
+                let suppressSheet = UIAlertController(title: nil, message: Strings.SuppressAlertsActionMessage, preferredStyle: .actionSheet)
+                suppressSheet.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .destructive, handler: { (action) in
```
Probably cleaner to make `suppressDialogues` take an empty `UIAlertAction`:

`func suppressDialogues(_: UIAlertAction) {`

Since then you can just pass it in without creating a new closure around it:

```
suppressSheet.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .destructive, handler: suppressDialogues))
```

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,60 @@ extension BrowserViewController: WKUIDelegate {
+                suppressSheet.addAction(UIAlertAction(title: Strings.SuppressAlertsActionTitle, style: .destructive, handler: { (action) in
+                    suppressDialogues()
+                }))
+                suppressSheet.addAction(UIAlertAction(title: Strings.CancelButtonTitle, style: .cancel, handler: { (action) in
```
If not using `action` here, should use `_`

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,60 @@ extension BrowserViewController: WKUIDelegate {
+            return
+        }
+        promptingTab.alertShownCount += 1
+        var suppressBlock: JSAlertInfo.SuppressHandler = ({[unowned self] suppress in
```
Remove leading and trailing `(`/`)`

e.g. `let someBlock = ({})` us unnecessary.

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,60 @@ extension BrowserViewController: WKUIDelegate {
+        }
+    }
+    
+    func handleAlert<T: JSAlertInfo>(webView: WKWebView, alert: inout T, completionHandler: @escaping (Any?) -> Void) {
```
All of the callers into this function are looking for a value passed into the `completionHandler`, but don't see any ever being passed. Am I missing something, or is this not actually needed?

danishjafri88 replied:
True, updated the code! Thanks

jhreis replied:
If we're just always passing in `nil`, why not just remove it and make it `() -> Void`?

danishjafri88 committed:
Made linter changes in JSAlert Code

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,58 @@ extension BrowserViewController: WKUIDelegate {
+                suppressSheet.addAction(UIAlertAction(title: Strings.CancelButtonTitle, style: .cancel, handler: { _ in
+                    completionHandler(nil)
+                }))
+                self.present(suppressSheet, animated: true, completion: nil)
```
`completion: nil` is not needed, if this is `nil` it can just be removed entirely.

jhreis commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2145,37 +2145,58 @@ extension BrowserViewController: WKUIDelegate {
-            promptingTab.queueJavascriptAlertPrompt(textInputAlert)
+            let controller = alert.alertController()
+            controller.delegate = self
+            present(controller, animated: true, completion: nil)
```
Can remove `completion: nil` (I know this exited before)

danishjafri88 committed:
Refactored Browser Prompt code

jhreis commented:
Looks good, waiting on sec review.

diracdeltas commented:
Assigning to Tom per https://github.com/brave/brave-ios/pull/770#issuecomment-457403888. I approve having this feature, haven't tested it out.

jumde commented:
Running into this crash while running this code:

```
2019-05-20 09:53:53.916186-0700 Client[376:28411] *** Terminating app due to uncaught exception 'NSGenericException', reason: 'Your application has presented a UIAlertController (<UIAlertController: 0x10f849c00>) of style UIAlertControllerStyleActionSheet from UINavigationController (<UINavigationController: 0x10680da00>). The modalPresentationStyle of a UIAlertController with this style is UIModalPresentationPopover. You must provide location information for this popover through the alert controller's popoverPresentationController. You must provide either a sourceView and sourceRect or a barButtonItem.  If this information is not known when you present the alert controller, you may provide it in the UIPopoverPresentationControllerDelegate method -prepareForPopoverPresentation.'
*** First throw call stack:
(0x18932f518 0x18850a9f8 0x1b5a35cb4 0x1b5a3e51c 0x1b5a3c068 0x1b60c065c 0x1b60af1ac 0x1b60dc01c 0x1892c089c 0x1892bb5c4 0x1892bbb40 0x1892bb354 0x18b4bb79c 0x1b60b51e8 0x102bb8728 0x188d818e0)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

with this code:
```
$ cat js.html
<script>
   while(true) {
     alert(1);
   }
</script>
```

on clicking supress alerts.

danishjafri88 committed:
Added action sheet support for iPad

danishjafri88 committed:
suppressing Javascript Alert functions

danishjafri88 commented:
> 'Your application has presented a UIAlertController (<UIAlertController: 0x10f849c00>) of style UIAlertControllerStyleActionSheet from UINavigationController (<UINavigationController: 0x10680da00>). The modalPresentationStyle of a UIAlertController with this style is UIModalPresentationPopover. You must provide location information for this popover through the alert controller's popoverPresentationController. You must provide either a sourceView and sourceRect or a barButtonItem.  If this information is not known when you present the alert controller, you may provide it in the UIPopoverPresentationControllerDelegate method -prepareForPopoverPresentation

This was caused as the alert was not handled for iPad. :( 
Fixed now.

Added another script to suppress js alerts to keep ui further responsive.

danishjafri88 committed:
Updated the script to handle frames also

danishjafri88 commented:
@jumde I have updated the script to handle iframes too. Please give it a try once.
